service/history/shardContext.go

```diff
@@ -856,24 +868,23 @@ func (s *shardContextImpl) getRangeID() int64 {
 	return s.shardInfo.RangeID
 }
 
+func (s *shardContextImpl) isClosed() bool {
+	return atomic.LoadInt32(&s.closed) != 0
+}
+
 func (s *shardContextImpl) closeShard() {
-	if s.isClosed {
+	if !atomic.CompareAndSwapInt32(&s.closed, 0, 1) {
 		return
 	}
 
-	s.isClosed = true
-
-	go s.shardItem.stopEngine()
+	go func() {
+		s.closeCallback(s.shardID, s.shardItem)
+		s.shardItem.stopEngine()
```

Comment:
does closeCallback call stopEngine?

Reply:
Yes - forgot to remove this additional call to stopEngine. Will remove it.

Reply:
done